In [34]:
import pandas as pd

# 8.1 Parsing Unix timestamps

It's not obvious how to deal with Unix timestamps in pandas -- it took me quite a while to figure this out. The file we're using here is a popularity-contest file I found on my system at `/var/log/popularity-contest`.

Here's an [explanation of how this file works](http://popcon.ubuntu.com/README).

I'm going to hope that nothing in it is sensitive :)

### Poking around with this data set 

In [23]:
popcon = pd.read_csv('../data/popularity-contest', sep=' ', )
popcon

POPULARITY-CONTEST-0  TIME:1387295813  \
0                   1387295797       1367633260   
1                   1387295796       1354370480   
2                   1387295743       1354341275   
3                   1387295743       1387224204   
4                   1387295742       1354341253   
...                        ...              ...   
2893                         0                0   
2894                         0                0   
2895                         0                0   
2896                         0                0   
2897  END-POPULARITY-CONTEST-0  TIME:1387295813   

         ID:d9bdc557ae8941c19e95da1c5da786bd  \
0                                  perl-base   
1                                      login   
2                                 libtalloc2   
3                               libwbclient0   
4                                libselinux1   
...                                      ...   
2893                        notify-osd-icons   
2894                       python-apt-common   
2895  libindicator-messages-status-provider1   
2896                             libusbmuxd1   
2897                                     NaN   

                                        ARCH:amd64 POPCONVER:1.53ubuntu1  
0                                    /usr/bin/perl                   NaN  
1                                          /bin/su                   NaN  
2     /usr/lib/x86_64-linux-gnu/libtalloc.so.2.0.7                   NaN  
3       /usr/lib/x86_64-linux-gnu/libwbclient.so.0        <RECENT-CTIME>  
4            /lib/x86_64-linux-gnu/libselinux.so.1                   NaN  
...                                            ...                   ...  
2893                                     <NOFILES>                   NaN  
2894                                     <NOFILES>                   NaN  
2895                                     <NOFILES>                   NaN  
2896                                     <NOFILES>                   NaN  
2897                                           NaN                   NaN  

[2898 rows x 5 columns]

It has both a header (which read_csv used) and a trailer which marks the end of the dataset. We can remove this with the slicer [:-1]. Check its shape and tail

In [35]:
popcon = popcon[:-1]
popcon.shape

(2896, 5)

In [25]:
# Trimmed tail
popcon.tail()

POPULARITY-CONTEST-0 TIME:1387295813  \
2892                    0               0   
2893                    0               0   
2894                    0               0   
2895                    0               0   
2896                    0               0   

         ID:d9bdc557ae8941c19e95da1c5da786bd ARCH:amd64 POPCONVER:1.53ubuntu1  
2892                         libreadline-dev  <NOFILES>                   NaN  
2893                        notify-osd-icons  <NOFILES>                   NaN  
2894                       python-apt-common  <NOFILES>                   NaN  
2895  libindicator-messages-status-provider1  <NOFILES>                   NaN  
2896                             libusbmuxd1  <NOFILES>                   NaN

### Put it together
Read it, remove the last row, and rename the columns

In [28]:
# Read it, and remove the last row
popcon = pd.read_csv('../data/popularity-contest', sep=' ', )[:-1]
popcon.columns = ['atime', 'ctime', 'package-name', 'mru-program', 'tag']

The colums are the access time, created time, package name, recently used program, and a tag

In [36]:
popcon[:5]

atime       ctime  package-name  \
0  1387295797  1367633260     perl-base   
1  1387295796  1354370480         login   
2  1387295743  1354341275    libtalloc2   
3  1387295743  1387224204  libwbclient0   
4  1387295742  1354341253   libselinux1   

                                    mru-program             tag  
0                                 /usr/bin/perl             NaN  
1                                       /bin/su             NaN  
2  /usr/lib/x86_64-linux-gnu/libtalloc.so.2.0.7             NaN  
3    /usr/lib/x86_64-linux-gnu/libwbclient.so.0  <RECENT-CTIME>  
4         /lib/x86_64-linux-gnu/libselinux.so.1             NaN

The magical part about parsing timestamps in pandas is that numpy datetimes are already stored as Unix timestamps. So all we need to do is tell pandas that these integers are actually datetimes -- it doesn't need to do any conversion at all.

We need to convert these to ints to start:

In [41]:
popcon['atime'] = popcon['atime'].astype(int)
popcon['ctime'] = popcon['ctime'].astype(int)

Every numpy array and pandas series has a dtype -- this is usually `int64`, `float64`, or `object`. Some of the time types available are `datetime64[s]`, `datetime64[ms]`, and `datetime64[us]`. There are also `timedelta` types, similarly.

We can use the `pd.to_datetime` function to convert our integer timestamps into datetimes. This is a constant-time operation -- we're not actually changing any of the data, just how pandas thinks about it.

In [43]:
popcon['atime'] = pd.to_datetime(popcon['atime'], unit='s')
popcon['ctime'] = pd.to_datetime(popcon['ctime'], unit='s')

If we look at the dtype now, it's `<M8[ns]`. As far as I can tell `M8` is secret code for `datetime64`.

In [50]:
# popcon['atime'].dtype
popcon[['atime', 'ctime']].dtypes

atime    datetime64[ns]
ctime    datetime64[ns]
dtype: object

So now we can look at our `atime` and `ctime` as dates!

In [51]:
popcon[:5]

atime               ctime  package-name  \
0 2013-12-17 15:56:37 2013-05-04 02:07:40     perl-base   
1 2013-12-17 15:56:36 2012-12-01 14:01:20         login   
2 2013-12-17 15:55:43 2012-12-01 05:54:35    libtalloc2   
3 2013-12-17 15:55:43 2013-12-16 20:03:24  libwbclient0   
4 2013-12-17 15:55:42 2012-12-01 05:54:13   libselinux1   

                                    mru-program             tag  
0                                 /usr/bin/perl             NaN  
1                                       /bin/su             NaN  
2  /usr/lib/x86_64-linux-gnu/libtalloc.so.2.0.7             NaN  
3    /usr/lib/x86_64-linux-gnu/libwbclient.so.0  <RECENT-CTIME>  
4         /lib/x86_64-linux-gnu/libselinux.so.1             NaN

Now suppose we want to look at all packages that aren't libraries.

First, I want to get rid of everything with timestamp 0. Notice how we can just use a string in this comparison, even though it's actually a timestamp on the inside? That is because pandas is amazing.

In [52]:
popcon = popcon[popcon['atime'] > '1970-01-01']

Now we can use pandas' magical string abilities to just look at rows where the package name doesn't contain 'lib'.

In [53]:
nonlibraries = popcon[~popcon['package-name'].str.contains('lib')]

In [55]:
nonlibraries.sort_values('ctime', ascending=False)[:10]

atime               ctime               package-name  \
57  2013-12-17 04:55:39 2013-12-17 04:55:42                        ddd   
450 2013-12-16 20:03:20 2013-12-16 20:05:13                     nodejs   
454 2013-12-16 20:03:20 2013-12-16 20:05:04  switchboard-plug-keyboard   
445 2013-12-16 20:03:20 2013-12-16 20:05:04      thunderbird-locale-en   
396 2013-12-16 20:08:27 2013-12-16 20:05:03            software-center   
449 2013-12-16 20:03:20 2013-12-16 20:05:00           samba-common-bin   
397 2013-12-16 20:08:25 2013-12-16 20:04:59      postgresql-client-9.1   
398 2013-12-16 20:08:23 2013-12-16 20:04:58             postgresql-9.1   
452 2013-12-16 20:03:20 2013-12-16 20:04:55                   php5-dev   
440 2013-12-16 20:03:20 2013-12-16 20:04:54                   php-pear   

                                           mru-program             tag  
57                                        /usr/bin/ddd  <RECENT-CTIME>  
450                                       /usr/bin/npm  <RECENT-CTIME>  
454       /usr/lib/plugs/pantheon/keyboard/options.txt  <RECENT-CTIME>  
445  /usr/lib/thunderbird-addons/extensions/langpac...  <RECENT-CTIME>  
396                   /usr/sbin/update-software-center  <RECENT-CTIME>  
449                                /usr/bin/net.samba3  <RECENT-CTIME>  
397                   /usr/lib/postgresql/9.1/bin/psql  <RECENT-CTIME>  
398             /usr/lib/postgresql/9.1/bin/postmaster  <RECENT-CTIME>  
452                  /usr/include/php5/main/snprintf.h  <RECENT-CTIME>  
440                        /usr/share/php/XML/Util.php  <RECENT-CTIME>

Okay, cool, it says that I I installed ddd recently. And postgresql! I remember installing those things. Neat.

The whole message here is that if you have a timestamp in seconds or milliseconds or nanoseconds, then you can just "cast" it to a `'datetime64[the-right-thing]'` and pandas/numpy will take care of the rest.

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  